src/snowflake/snowpark/mock/connection.py

```diff
@@ -539,19 +563,33 @@ def get_result_query_id(self, plan: SnowflakePlan, **kwargs) -> str:
         return result_set["sfqid"]
 
 
-def _fix_pandas_df_integer(
-    pd_df: "pandas.DataFrame", results_cursor: SnowflakeCursor
-) -> "pandas.DataFrame":
-    for column_metadata, pandas_dtype, pandas_col_name in zip(
-        results_cursor.description, pd_df.dtypes, pd_df.columns
-    ):
+def _fix_pandas_df_integer(table_res: TableEmulator) -> "pandas.DataFrame":
+    pd_df = pd.DataFrame()
+    for col_name in table_res.columns:
+        col_sf_type = table_res.sf_types[col_name]
+        pd_df_col_name = unquote_if_quoted(col_name)
         if (
-            FIELD_ID_TO_NAME.get(column_metadata.type_code) == "FIXED"
-            and column_metadata.precision is not None
-            and column_metadata.scale == 0
-            and not str(pandas_dtype).startswith("int")
+            isinstance(col_sf_type.datatype, DecimalType)
+            and col_sf_type.datatype.precision is not None
+            and col_sf_type.datatype.scale == 0
         ):
-            pd_df[pandas_col_name] = pandas.to_numeric(
```

Comment:
Question: why can't we use the same code here?

Reply:
pandas.to_numeric doesn't performa the type cast